README
Building the OTRS package
=========================

1. Use opm.pl to create the package, thus:

[otrs home]/bin/opm.pl -a build -p /path/to/OpenNMS.sopm -d /path/to/

for me, with otrs in /Users/user/Documents/otrs/otrs and the otrs-opennms code checked out to /Users/user/svn/otrs-opennms:

jsartin-mac:~ user$ /Users/user/Documents/otrs/otrs/bin/opm.pl -a build -p /Users/user/svn/otrs-opennms/otrs/OpenNMS.sopm -d /Users/user/svn/otrs-opennms/otrs/
Writing /tmp/OpenNMS-0.1.0.opm
jsartin-mac:~ user$


Add the package to otrs
=======================

1. At the webUI for OTRS, go to Admin->Package Manager.
2. Add the package (/tmp/otrs-opennms-0.1.0.opm).
3. Hit "continue" at the dire warning page.

Add the SOAP user to OTRS
=========================

1. At the webUI for OTRS, go to Admin->SysConfig.
2. Show the Group for "Framework".
3. Select Core::SOAP
4. Set the SOAP::User and SOAP::Password (the sample java code expects opennms/opennms). The opennms package uses the same authentication technique as the existing SOAP handle.

Check that you have the required Perl Modules installed
=======================================================

Ensure you have the SOAP::Lite, and SOAP::DateTime CPAN modules installed. If you have used the existing OTRS SOAP handle, you will probably already have SOAP::Lite.

Compile the Java and run the tests
==================================

You'll need maven for this. At the root of the otrs-opennms package that you downloaded from the OpenNMS repo, run:

jsartin-mac:~/Documents/otrs/otrs-opennms user$ mvn test

This will compile the code and run the tests (there are 7 of them). The tests expect to see your OTRS server at http://localhost/otrs/. You should now see some tickets in your otrs raw queue.

Hope that this has been of some use to you.

Jonathan Sartin (jonathan@opennms.org)


CHANGELOG
1.0.0 to 1.1.0
----------------------------------------------------------------------
- Added German installation into
- Now works with OTRS 2.4

  Patches gratefully recieved from Martin Edenhofer.

1.1.0 to 1.2.0
----------------------------------------------------------------------
- Removed dependency on the hard-to-get SOAP::DateTime. Now calls
  Date::Manip directly.


otrs/Kernel/System/WebService.pm
# This file is part of the OpenNMS(R) Application.
#
# OpenNMS(R) is Copyright (C) 2002-2008 The OpenNMS Group, Inc. All rights
# reserved.  OpenNMS(R) is a derivative work, containing both original code,
# included code and modified code that was published under the GNU General
# Public License. Copyrights for modified and included code are below.
#
# OpenNMS(R) is a registered trademark of The OpenNMS Group, Inc.
#
# Copyright (C) Jonathan Sartin (Jonathan@opennms.org)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
#
# For more information contact:
# OpenNMS Licensing <license@opennms.org>
# http://www.opennms.org/
# http://www.opennms.com/

# Kernel/System/WebService.pm - Super Class for Ticket/Article interface for OpenNMS Integration


package Kernel::System::WebService;

use strict;
use warnings;

use SOAP::Lite;
use Kernel::Config;
use Kernel::System::Encode;
use Kernel::System::Log;
use Kernel::System::DB;
use Kernel::System::PID;
use Kernel::System::Main;
use Kernel::System::Time;
use Kernel::System::User;
use Kernel::System::Group;
use Kernel::System::Queue;
use Kernel::System::Ticket;

use base qw( SOAP::Server::Parameters );

use vars qw($VERSION);
$VERSION = qw($Revision: 18 $) [1];

sub new {

	my $Class = shift;
	my $Self = {};
	my $som = pop();

	# common objects - create as a hashref

	$Self->{CommonObject} = {};
	$Self->{CommonObject}->{ConfigObject} = Kernel::Config->new();
	$Self->{CommonObject}->{EncodeObject} =
		Kernel::System::Encode->new(%{$Self->{CommonObject}});
	$Self->{CommonObject}->{LogObject}    = Kernel::System::Log->new(
    	LogPrefix => 'OTRS-SOAP',
    	%{$Self->{CommonObject}},
	);
	$Self->{CommonObject}->{MainObject}   = 
		Kernel::System::Main->new(%{$Self->{CommonObject}});
	$Self->{CommonObject}->{DBObject}     = 
		Kernel::System::DB->new(%{$Self->{CommonObject}});
	$Self->{CommonObject}->{PIDObject}    = 
		Kernel::System::PID->new(%{$Self->{CommonObject}});
	$Self->{CommonObject}->{TimeObject}   = 
		Kernel::System::Time->new(%{$Self->{CommonObject}});
	$Self->{CommonObject}->{UserObject}   = 
		Kernel::System::User->new(%{$Self->{CommonObject}});
	$Self->{CommonObject}->{GroupObject}  = 
		Kernel::System::Group->new(%{$Self->{CommonObject}});
	$Self->{CommonObject}->{QueueObject}  = 
		Kernel::System::Queue->new(%{$Self->{CommonObject}});
	$Self->{CommonObject}->{TicketObject} = 
		Kernel::System::Ticket->new(%{$Self->{CommonObject}});
	
	my $header = $som->header();
	my $RequiredUser = $Self->{CommonObject}->{ConfigObject}->Get('SOAP::User');
 	my $RequiredPassword = $Self->{CommonObject}->{ConfigObject}->Get('SOAP::Password');

 	$Self->{CommonObject}->{LogObject}->Log(
		Priority => 'debug',
		Message  => "user: (required) $RequiredUser - (request) $header->{request_header}->{User}",
	);
 
    if ( !defined $RequiredUser || !length( $RequiredUser )
        || !defined $RequiredPassword || !length( $RequiredPassword )
    ) {
        $Self->{CommonObject}->{LogObject}->Log(
            Priority => 'notice',
            Message  => "SOAP::User or SOAP::Password is empty, SOAP access denied!",
        );
        die SOAP::Fault
        	->faultcode('Server.RequestError')
        	->faultstring("Authentication Failure");
    }
    if ( $header->{request_header}->{User} ne $RequiredUser || $header->{request_header}->{Pass} ne $RequiredPassword ) {
        $Self->{CommonObject}->{LogObject}->Log(
            Priority => 'notice',
            Message  => "Auth for user $header->{request_header}->{User} failed!",
        );
        die SOAP::Fault
        	->faultcode('Server.RequestError')
        	->faultstring("Authentication Failure");
    }
    
    bless($Self, $Class);
	
	return $Self;
  
}


1;


src/main/java/org/opennms/integration/otrs/ticketclient/TicketClient.java
package org.opennms.integration.otrs.ticketclient;

import org.opennms.integration.otrs.ticketservice.*;

public class TicketClient {

	public static void main(String[] args) throws Exception {

		TicketServiceLocator service = new TicketServiceLocator();

		/*
		 * uncomment the following to change the endpoint to match your OTRS
		 * server
		 * 
		 * service.setTicketServicePortEndpointAddress(new
		 * java.lang.String("http://myhost:8080/otrs/soap.pl"));
		 * 
		 */

		TicketServicePort_PortType port = service.getTicketServicePort();

		// modify pass and user to suit those defined for your OTRS soap user.

		Credentials creds = new Credentials();

		creds.setPass("pass1");
		creds.setUser("user1");

		TicketCore newTicket = new TicketCore();

		/*
		 * TicketCore needs: [Queue|QueueID] [Priority|PriorityID]
		 * [State|StateID] [Lock|LockID] OwnerID UserID Title
		 */

		newTicket.setQueue("Raw");
		newTicket.setPriority("3 normal");
		newTicket.setState("new");
		newTicket.setLock("unlock");
		newTicket.setOwnerID(1);
		newTicket.setUserID(1);
		newTicket.setTitle("OpenNMS Integration Test Ticket");

		TicketIDAndNumber idAndNumber = port.ticketCreate(newTicket, creds);

		System.out.println("Created Ticket with ID: "
				+ idAndNumber.getTicketID() + " and number: "
				+ idAndNumber.getTicketNumber());

		ArticleCore article1 = new ArticleCore();

		/*
		 * ArticleCore needs: TicketID [ArticleType|ArticleTypeID]
		 * [SenderType|SenderTypeID] From Subject ContentType HistoryType
		 * HistoryCOmment Body UserID
		 */

		article1.setTicketID(idAndNumber.getTicketID());
		article1.setArticleType("note-external");
		article1.setFrom("jonathan@opennms.org");
		article1.setSenderType("agent");
		article1.setSubject("first article");
		article1.setBody("body text for first article");
		article1.setContentType("text/plain; charset=ISO-8859-15");
		article1.setHistoryComment("updated by OpenNMS");
		article1.setHistoryType("OwnerUpdate");
		article1.setUserID(1);

		int article1ID = port.articleCreate(article1, creds);

		System.out.println("Created Article with ID: " + article1ID
				+ " for Ticket ID: " + idAndNumber.getTicketID());

		ArticleCore article2 = new ArticleCore();

		article2.setTicketID(idAndNumber.getTicketID());
		article2.setArticleType("note-external");
		article2.setFrom("jonathan@opennms.org");
		article2.setSenderType("agent");
		article2.setSubject("second article");
		article2.setBody("body text for second article");
		article2.setContentType("text/plain; charset=ISO-8859-15");
		article2.setHistoryComment("updated by OpenNMS");
		article2.setHistoryType("OwnerUpdate");
		article2.setUserID(1);

		int article2ID = port.articleCreate(article2, creds);

		System.out.println("Created Article with ID: " + article2ID
				+ " for Ticket ID: " + idAndNumber.getTicketID());

		// Now play it back ...

		/*
		 * could use:
		 * 
		 * port.ticketGetById(idAndNumber.getTicketId(), creds);
		 * 
		 * instead ...
		 */

		Ticket ticket = port.ticketGetByNumber(idAndNumber.getTicketNumber(),
				creds);

		System.out.println("Ticket with ID: " + ticket.getTicketID()
				+ " and number " + ticket.getTicketNumber());
		System.out.println("Title:    " + ticket.getTitle());

		/*
		 * could use:
		 * 
		 * port.articleGetAllByTicketId(idAndNumber.getTicketId(), creds);
		 * 
		 * instead ...
		 */

		Article[] articlesForTicket = port.articleGetAllByTicketNumber(
				idAndNumber.getTicketNumber(), creds);

		for (Article article : articlesForTicket) {
			System.out.println("Ticket   " + idAndNumber.getTicketID()
					+ " contains article " + article.getArticleID() + " with:");
			System.out.println("Type:    " + article.getArticleType());
			System.out.println("From:    " + article.getFrom());
			System.out.println("Subject: " + article.getSubject());
			System.out.println("Body:    " + article.getBody());
		}
	}
}


otrs/Kernel/System/TicketService.pm
# This file is part of the OpenNMS(R) Application.
#
# OpenNMS(R) is Copyright (C) 2002-2008 The OpenNMS Group, Inc. All rights
# reserved.  OpenNMS(R) is a derivative work, containing both original code,
# included code and modified code that was published under the GNU General
# Public License. Copyrights for modified and included code are below.
#
# OpenNMS(R) is a registered trademark of The OpenNMS Group, Inc.
#
# Copyright (C) Jonathan Sartin (Jonathan@opennms.org)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
#
# For more information contact:
# OpenNMS Licensing <license@opennms.org>
# http://www.opennms.org/
# http://www.opennms.com/

package Kernel::System::TicketService;

@ISA = ("Kernel::System::WebService");

use strict;
use warnings;

use Kernel::System::WebService;
use Date::Manip;
use integer;
use constant URI => "http://opennms.org/integration/otrs/TicketService";

use vars qw($VERSION);
$VERSION = qw($Revision: 19 $) [1];

sub new {
 
    my $class = shift;
    
    return $class if ref($class);
    
    my $Self = $class->SUPER::new(@_);
 
    return $Self;

}

sub TicketGetByID() {

	my $Self = shift->new(@_);
	my $TicketID = shift;

	my %Ticket = $Self->{CommonObject}->{TicketObject}->TicketGet(
		TicketID => $TicketID
	);

	unless (%Ticket) {
		$Self->{CommonObject}->{LogObject}->Log(
			Priority => 'error',
			Message => "No such ticket: $TicketID");
    	die SOAP::Fault
       		->faultcode('Server.RequestError')
        	->faultstring("No such ticket: $TicketID");
	}
	
	my $Parsed = ParseDate( $Ticket{Created} );
	unless($Parsed) {
               	$Self->{CommonObject}->{LogObject}->Log(
                        Priority => 'error',
                        Message => "Could not parse creation date");
        	die SOAP::Fault
               		->faultcode('Server.RequestError')
                	->faultstring("Could not parse creation date");
        }

  	my $Created =  UnixDate($Parsed,"%Y-%m-%dT%H:%M:%S");

	my @TicketResponse = 
	(
		SOAP::Data->name("TicketID" => $Ticket{TicketID})->type("int"),
		SOAP::Data->name("TicketNumber" => $Ticket{TicketNumber})->type("long"),
		SOAP::Data->name("Title" => $Ticket{Title})->type("string"),
		SOAP::Data->name("Age" => $Ticket{Age})->type("int"),
		SOAP::Data->name("State" => $Ticket{State})->type("string"),
		SOAP::Data->name("StateID" => $Ticket{StateID})->type("int"),
		SOAP::Data->name("StateType" => $Ticket{State})->type("string"),
		SOAP::Data->name("Priority" => $Ticket{Priority})->type("string"),
		SOAP::Data->name("PriorityID" => $Ticket{PriorityID})->type("int"),
		SOAP::Data->name("Lock" => $Ticket{Lock})->type("string"),
		SOAP::Data->name("LockID" => $Ticket{LockID})->type("int"),
		SOAP::Data->name("UnlockTimeout" => $Ticket{UnlockTimeout})->type("int"),
		SOAP::Data->name("Queue" => $Ticket{Queue})->type("string"),
		SOAP::Data->name("QueueID" => $Ticket{QueueID})->type("int"),
		SOAP::Data->name("GroupID" => $Ticket{GroupID})->type("int"),
		SOAP::Data->name("CustomerID" => $Ticket{CustomerID})->type("string"),
		SOAP::Data->name("CustomerUserID" => $Ticket{CustomerUserID})->type("string"),
		SOAP::Data->name("Owner" => $Ticket{Owner})->type("string"),
		SOAP::Data->name("OwnerID" => $Ticket{OwnerID})->type("int"),
		SOAP::Data->name("Responsible" => $Ticket{Responsible})->type("string"),
		SOAP::Data->name("ResponsibleID" => $Ticket{ResponsibleID})->type("int"),
		SOAP::Data->name("Type" => $Ticket{Type})->type("string"),
		SOAP::Data->name("TypeID" => $Ticket{TypeID})->type("int"),
		SOAP::Data->name("SLAID" => $Ticket{SLAID})->type("int"),
		SOAP::Data->name("ServiceID" => $Ticket{ServiceID})->type("int"),
		SOAP::Data->name("Created" => $Created)->type("dateTime"),
		SOAP::Data->name("CreateTimeUnix" => $Ticket{CreateTimeUnix})->type("long"),
		SOAP::Data->name("UntilTime" => $Ticket{UntilTime})->type("long"),
		SOAP::Data->name("EscalationStartTime" => $Ticket{EscalationStartTime})->type("long"),
		SOAP::Data->name("EscalationResponseTime" => $Ticket{EscalationResponseTime})->type("long"),
		SOAP::Data->name("EscalationSolutionTime" => $Ticket{EscalationSolutionTime})->type("long"),
	);
    
    return SOAP::Data->name( "Ticket" )
			->attr( {"xmlns:tns" => URI } )
			->type( "tns:Ticket" )
			->value( \SOAP::Data->value(@TicketResponse) );
	
}

sub TicketGetByNumber() {

	my $Self = shift->new(@_);
	my $TicketNumber = shift();
		
	my $TicketID = $Self->{CommonObject}->{TicketObject}->TicketIDLookup(TicketNumber => $TicketNumber);
    
    unless ($TicketID) {
		$Self->{CommonObject}->{LogObject}->Log( Priority => 'error', 
			Message => "No TicketID for TicketNumber $TicketNumber" );
       	die SOAP::Fault
       		->faultcode('Server.RequestError')
        	->faultstring("No TicketID for TicketNumber $TicketNumber");
	}
    
    return $Self->TicketGetByID($TicketID);
    
}

sub GetById() {
	
	my $Self = shift->new(@_);
	my $TicketID = shift();
		
    my $Ticket =  $Self->TicketGetByID($TicketID);
    my $Articles =  $Self->ArticleGetAllByTicketID($TicketID);
    
    my @GetByIdResponse = 
	(
		SOAP::Data->value($Ticket),
		SOAP::Data->value($Articles)
 	);
 	
    return SOAP::Data->name( "ticketWithArticles" )
			->attr( {"xmlns:tns" => URI } )
			->type( "tns:TicketWithArticles" )
			->value( \SOAP::Data->value(@GetByIdResponse) );
	
}

sub GetByNumber() {
	
	my $Self = shift->new(@_);
	my $TicketNumber = shift();
		
	my $TicketID = $Self->{CommonObject}->{TicketObject}->TicketIDLookup(TicketNumber => $TicketNumber);
    
    unless ($TicketID) {
		$Self->{CommonObject}->{LogObject}->Log( Priority => 'error', 
			Message => "No TicketID for TicketNumber $TicketNumber" );
       	die SOAP::Fault
       		->faultcode('Server.RequestError')
        	->faultstring("No TicketID for TicketNumber $TicketNumber");
	};
    
    my $Ticket =  $Self->TicketGetByID($TicketID);
    my $Articles =  $Self->ArticleGetAllByTicketID($TicketID);
    
    my @GetByNumberResponse = 
	(
		SOAP::Data->value($Ticket),
		SOAP::Data->value($Articles)
 	);
 	
    return SOAP::Data->name( "ticketWithArticles" )
			->attr( {"xmlns:tns" => URI } )
			->type( "tns:TicketWithArticles" )
			->value( \SOAP::Data->value(@GetByNumberResponse) );
	
}


sub TicketCreate() {
	
	my $Self = shift->new(@_);
	my $TicketReq = shift;
	my $UserID = shift;
	my %Ticket;
	
    my @AlternateField = ('Queue', 'Priority', 'State', 'Lock');
    
    foreach (@AlternateField) {
		if ( $TicketReq->{$_.'ID'} ) {
			$Ticket{$_.'ID'} = $TicketReq->{$_.'ID'};
		} elsif ($TicketReq->{$_}) {
			$Ticket{$_} = $TicketReq->{$_};
		} else {
			die SOAP::Fault
       			->faultcode('Server.RequestError')
       			->faultstring('Require '.$_.' or '.$_.'ID');
		}
    }
    
    my @RequiredField = ('Title', 'OwnerID');
    
    foreach (@RequiredField) {
		if ( $TicketReq->{$_} ) {
			$Ticket{$_} = $TicketReq->{$_};
		} else {
			die SOAP::Fault
       			->faultcode('Server.RequestError')
       			->faultstring("Require $_ ");
		}
    }
    
    # Need UserID or User, but have to look up UserID if given User.
    
    if ( $TicketReq->{UserID} ) {
			$Ticket{UserID} = $TicketReq->{UserID};
		} elsif ($TicketReq->{User}) {
			$Ticket{UserID} = $Self->GetUserIDForUser( \$TicketReq->{User} );
		} else {
			die SOAP::Fault
       			->faultcode('Server.RequestError')
       			->faultstring('Require UserID or User');
	}
	

	my $TicketID = $Self->{CommonObject}->{TicketObject}->TicketCreate(%Ticket);
    
    unless ($TicketID) {
    	die SOAP::Fault
    		->faultcode('Server.RequestError')
        	->faultstring("Unable to create Ticket");
    }
    
    my $TicketNumber = $Self->{CommonObject}->{TicketObject}->TicketNumberLookup(TicketID => $TicketID);
    
    my @TicketIDAndNumber = 
	(
		SOAP::Data->name("TicketID" => $TicketID)->type("int"),
		SOAP::Data->name("TicketNumber" => $TicketNumber)->type("long"),
	);


    return SOAP::Data->name( "TicketIDAndNumber" )
			->attr( {"xmlns:tns" => URI } )
			->type( "tns:TicketIDAndNumber" )
			->value( \SOAP::Data->value(@TicketIDAndNumber) );

}

sub TicketStateUpdate() {

	my $Self = shift->new(@_);
	my $TicketStateUpdateReq = shift();
	my $UserID;
	my $TicketID;
		
	# Need TicketID or TicketNumber, make sure we've got one or the other. 
	# Look up the TicketID from the TicketNumber if that's what was presented.
	
	if ($TicketStateUpdateReq->{TicketNumber}) {
		$TicketID = $Self->{CommonObject}->{TicketObject}->TicketIDLookup(TicketNumber => $TicketStateUpdateReq->{TicketNumber});
	} elsif ($TicketStateUpdateReq->{TicketID}) {
		$TicketID = $TicketStateUpdateReq->{TicketID};
	} else {
		$Self->{CommonObject}->{LogObject}->Log(
			Priority => 'error',
			Message => "Require TicketID or TicketNumber");
		die SOAP::Fault
       		->faultcode('Server.RequestError')
        	->faultstring("Require TicketID or TicketNumber");
	};
	
	# Check the TicketID is valid
	
	unless ($Self->{CommonObject}->{TicketObject}->TicketGet(TicketID => $TicketID)) {
		$Self->{CommonObject}->{LogObject}->Log(
			Priority => 'error',
			Message => "No such ticket: $TicketID");
    	die SOAP::Fault
       		->faultcode('Server.RequestError')
        	->faultstring("No such ticket: $TicketID");
	}
	
	# Need UserID or User, but have to look up UserID if given User.
	
	if ( $TicketStateUpdateReq->{UserID} ) {
			$UserID = $TicketStateUpdateReq->{UserID};
		} elsif ($TicketStateUpdateReq->{User}) {
			$UserID = $Self->GetUserIDForUser( \$TicketStateUpdateReq->{User});
		} else {
			$Self->{CommonObject}->{LogObject}->Log(
				Priority => 'error',
				Message => "Require UserID or User");
			die SOAP::Fault
       			->faultcode('Server.RequestError')
       			->faultstring('Require UserID or User');
	}
	
	# Update, using ticketID and UserID with State or StatID depending on
	# which was offered in the request.
    
    if ( $TicketStateUpdateReq->{State} ) {
		$Self->{CommonObject}->{TicketObject}->StateSet(TicketID => $TicketID,
												UserID => $UserID,
												State => $TicketStateUpdateReq->{State});
		$Self->{CommonObject}->{LogObject}->Log(
				Priority => 'debug',
				Message => "Updated ticket $TicketStateUpdateReq->{TicketNumber} state to $TicketStateUpdateReq->{State}");
    	} elsif  ( $TicketStateUpdateReq->{StateID} ) {
    		$Self->{CommonObject}->{TicketObject}->StateSet(TicketID => $TicketID,
													UserID => $UserID,
													StateID => $TicketStateUpdateReq->{StateID});
    	} else {
    		$Self->{CommonObject}->{LogObject}->Log(
				Priority => 'error',
				Message => "Require StateID or State");
    		die SOAP::Fault
       			->faultcode('Server.RequestError')
       			->faultstring('Require StateID or State');
    		
    	}
}

sub ArticleCreate() {
	
	my $Self = shift->new(@_);
	my $ArticleReq = shift;
	my %Article;
	my $TicketID;
	
	# Better have a TicketID - check that first
	
	if ($ArticleReq->{TicketNumber}) {
		$TicketID = $Self->{CommonObject}->{TicketObject}->TicketIDLookup(TicketNumber => $ArticleReq->{TicketNumber});
	} elsif ($ArticleReq->{TicketID}) {
		$TicketID = $ArticleReq->{TicketID};
	} else {
		$Self->{CommonObject}->{LogObject}->Log(
			Priority => 'error',
			Message => "Require TicketID or TicketNumber");
		die SOAP::Fault
       		->faultcode('Server.RequestError')
        	->faultstring("Require TicketID or TicketNumber");
	};
	
	unless ($Self->{CommonObject}->{TicketObject}->TicketGet(TicketID => $TicketID)) {
		$Self->{CommonObject}->{LogObject}->Log(
			Priority => 'error',
			Message => "No such ticket: $TicketID");
    	die SOAP::Fault
       		->faultcode('Server.RequestError')
        	->faultstring("No such ticket: $TicketID");
	}
	
	$Article{TicketID} = $TicketID;
	
	
	# (Need ArticleType or ArticleTypeID) and (SenderType or SenderTypeID)
	# prefer ID version
	
	my @AlternateField = ('ArticleType', 'SenderType');
    
    foreach (@AlternateField) {
		if ( defined( $ArticleReq->{$_.'ID'} )) {
			$Article{$_.'ID'} = $ArticleReq->{$_.'ID'};
		} elsif (defined( $ArticleReq->{$_} )) {
			$Article{$_} = $ArticleReq->{$_};
		} else {
			die SOAP::Fault
       			->faultcode('Server.RequestError')
       			->faultstring('Require '.$_.' or '.$_.'ID');
		}
    }
    
    # Need From Subject Body ContentType HistoryType HistoryComment
    
    my @RequiredField = ('From', 'Subject', 'Body', 
    					 'ContentType', 'HistoryType', 'HistoryComment');
    
    foreach (@RequiredField) {
		if ( defined( $ArticleReq->{$_} )) {
			$Article{$_} = $ArticleReq->{$_};
		} else {
			die SOAP::Fault
       			->faultcode('Server.RequestError')
       			->faultstring("Require $_ ");
		}
    }
    
    # Need UserID or User, but have to look up UserID if given User.
    
    if ( $ArticleReq->{UserID} ) {
			$Article{UserID} = $ArticleReq->{UserID};
		} elsif ($ArticleReq->{User}) {
			$Article{UserID} = $Self->GetUserIDForUser( \$ArticleReq->{User});
		} else {
			die SOAP::Fault
       			->faultcode('Server.RequestError')
       			->faultstring('Require UserID or User');
	}
    
    my $ArticleID =  $Self->{CommonObject}->{TicketObject}->ArticleCreate(%Article);
    
     unless ($ArticleID) {
    	die SOAP::Fault
    		->faultcode('Server.RequestError')
        	->faultstring("Unable to create Article");
    }
    
    $Self->{CommonObject}->{LogObject}->Log( Priority => 'debug', 
			Message => "Created Article $ArticleID" );
    
    return $ArticleID;
}

sub ArticleGetByID() {

	my $Self = shift->new(@_);
	my $ArticleID = shift;

	my %Article = $Self->{CommonObject}->{TicketObject}->ArticleGet(
		ArticleID => $ArticleID
	);
	
	unless (%Article) {
		$Self->{CommonObject}->{LogObject}->Log(
			Priority => 'error',
			Message => "No such article: $ArticleID");
    	die SOAP::Fault
       		->faultcode('Server.RequestError')
        	->faultstring("No such article: $ArticleID");
	}

	my $FormattedArticle = FormatArticle($Self, \%Article);

	return $FormattedArticle;

}

sub ArticleGetAllByTicketID() {

	my $Self = shift->new(@_);
	my $TicketID = shift;
	my @ArticleList = ();
	
	unless ($Self->{CommonObject}->{TicketObject}->TicketGet(TicketID => $TicketID)) {
		$Self->{CommonObject}->{LogObject}->Log(
			Priority => 'error',
			Message => "No such ticket: $TicketID");
    	die SOAP::Fault
       		->faultcode('Server.RequestError')
        	->faultstring("No such ticket: $TicketID");
	}
	
	my @Articles = $Self->{CommonObject}->{TicketObject}->ArticleGet(
		TicketID => $TicketID
	);
	

	unless (@Articles) {
		$Self->{CommonObject}->{LogObject}->Log(
			Priority => 'error', 
			Message => "No articles available for ticket: $TicketID" );
    	die SOAP::Fault
       		->faultcode('Server.RequestError')
        	->faultstring("No articles available for ticket $TicketID");
	}
	
	foreach my $Article ( @Articles ) {
		push( @ArticleList, FormatArticle($Self,$Article) )
	}
	
	return FormatTypedArray( "Articles",
							 "Article",
							 \@ArticleList,
							 "tns",
							 URI);

}

sub ArticleGetAllByTicketNumber() {

	my $Self = shift->new(@_);
	my $TicketNumber = shift();
		
	my $TicketID = $Self->{CommonObject}->{TicketObject}->TicketIDLookup(TicketNumber => $TicketNumber);
    
    unless ($TicketID) {
		$Self->{CommonObject}->{LogObject}->Log( Priority => 'error', 
			Message => "No TicketID for TicketNumber $TicketNumber" );
       	die SOAP::Fault
       		->faultcode('Server.RequestError')
        	->faultstring("No TicketID for TicketNumber $TicketNumber");
	}
    
    return $Self->ArticleGetAllByTicketID($TicketID);
    
}

sub FormatArticle(){

	my $Self = shift;
	my $Article = shift();

    my @Message = 
	(
		SOAP::Data->name("ArticleID" => $Article->{ArticleID})->type("int"),
		SOAP::Data->name("TicketID" => $Article->{TicketID})->type("int"),
		SOAP::Data->name("ArticleTypeID" => $Article->{ArticleTypeID})->type("int"),
		SOAP::Data->name("ArticleType" => $Article->{ArticleType})->type("string"),
		SOAP::Data->name("SenderTypeID" => $Article->{SenderTypeID})->type("int"),
		SOAP::Data->name("SenderType" => $Article->{SenderType})->type("string"),
		SOAP::Data->name("From" => $Article->{From})->type("string"),
		SOAP::Data->name("To" => $Article->{To})->type("string"),
		SOAP::Data->name("CC" => $Article->{CC})->type("string"),
		SOAP::Data->name("Subject" => $Article->{Subject})->type("string"),
		SOAP::Data->name("ContentType" => $Article->{ContentType})->type("string"),
		SOAP::Data->name("MessageID" => $Article->{MessageID})->type("string"),
		SOAP::Data->name("IncomingTime" => $Article->{IncomingTime})->type("long"),
		SOAP::Data->name("ContentPath" => $Article->{ContentPath})->type("string"),
		SOAP::Data->name("ValidID" => $Article->{ValidID})->type("int"),
		SOAP::Data->name("Body" => $Article->{Body})->type("string"),
	);


    return SOAP::Data->name( "Article" )
			->attr( {"xmlns:tns" => URI } )
			->type( "tns:Article" )
			->value( \SOAP::Data->value(@Message) );

}

sub FormatTypedArray
{
  my ( $name, $type, $array, $ns, $uri ) = @_;
  $ns ||= "xsd";
  my $count = scalar( @$array );
  my $attr = {"soapenc:arrayType" =>
              "$ns:${type}\[$count]"};
  $attr->{"xmlns:$ns"} = $uri if ( $uri );

  my $formatted_array = [];
  
  foreach my $item ( @$array )
  {
    push ( @$formatted_array,
           SOAP::Data->type($type)->value($item) );
  }
  
  return SOAP::Data->name( $name )
                   ->attr( $attr )
                   ->value( $formatted_array );
}

sub GetUserIDForUser(){

	my $Self = shift;
	my $User = shift();
	my $UserID;
	
	
	$UserID = $Self->{CommonObject}->{UserObject}->UserLookup( UserLogin => ${$User} );

	unless ($UserID) {
		$Self->{CommonObject}->{LogObject}->Log(
			Priority => 'error',
			Message => "no UserID for ${$User}");
		die SOAP::Fault
       			->faultcode('Server.RequestError')
       			->faultstring("no UserID for ${$User}");
	};

    return $UserID;
}

1;
